web/modules/weather_data/src/Service/WeatherAlertTrait.php

```diff
@@ -190,6 +198,259 @@ public function getAlerts($grid, $point, $self = false)
+        $comparisonEnd,
+    ) {
+        return $subjectStart <= $comparisonStart &&
+            $subjectEnd >= $comparisonEnd;
+    }
+
+    /**
+     * Compute the duration of the alert
+     *
+     * The duration is a function of the DateTime
+     * diff, but adjusted for the overall period
+     * range.
+     * Note that DateInterval is "special"
+     * about how it computes diffs
+     */
+    private function calculateAlertDuration($onsetTime, $endTime, $max)
+    {
+        $alertDiff = $endTime->diff($onsetTime, true);
+        $alertDuration = $this->dateDiffInHours($alertDiff);
+
+        // If there are leftover minutes, we add
+        // an hour for coverage purposes
+        if ($alertDiff->m) {
+            $alertDuration += 1;
+        }
+
+        return min($max, $alertDuration);
+    }
+
+    /**
+     * Find relevant alert info in a list of hourly periods
+     *
+     * Given a list of hourly forecast periods and an alert,
+     * attempt to find alert alignment info within any of
+     * those periods.
+     * Return false if nothing is found
+     */
+    private function getAlertInfoInPeriods(
+        $alert,
+        $alertOnset,
+        $alertEnd,
+        $periods,
+        $timezone,
+    ) {
+        foreach ($periods as $periodIndex => $period) {
+            $periodStartTime = self::turnToDate(
+                $period["timestamp"],
+                $timezone,
+            );
+            $periodEndTime = $periodStartTime->modify("+ 1 hour");
+
+            $onsetIsWithinPeriod = $this->dateTimeIsWithin(
+                $alertOnset,
+                $periodStartTime,
+                $periodEndTime,
+            );
+            $endIsWithinPeriod = $this->dateTimeIsWithin(
+                $alertEnd,
+                $periodStartTime,
+                $periodEndTime,
+            );
+
+            if ($onsetIsWithinPeriod) {
+                // Get the number of hours the alert is
+                // supposed to last
+                $alertDuration = $this->calculateAlertDuration(
+                    $periodStartTime,
+                    $alertEnd,
+                    count($periods) - $periodIndex,
+                );
+
+                return [
+                    "duration" => $alertDuration,
+                    "periodIndex" => $periodIndex,
+                    "alert" => $alert,
+                ];
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Given an alert object, compute the endTime.
+     *
+     * Note that we select the `ends` field first.
+     * If null, we try the `expires` field.
+     * If both are null, return false.
+     */
+    private function getEndTimeForAlert($alert)
+    {
+        // We need to determine the correct ends field.
+        // If there is no value for the ends, then we use
+        // expires. If there is in that case no value for expires,
+        // then we do nothing and continue to the next alert,
+        // ignoring this one.
```

Comment:
@shadkeene @coreypieper @colinmurphy01 Another product question here, I think: should we have a ticket to determine what to do with alerts in the hourly forecast table when they don't have an `end` field but do have an `expires` field? My understanding is such an alert continues indefinitely but the expiration is there to prompt forecasters to update it; is that right?

Reply:
We should be using `eventEndingTime` (over `end`). That is the field in CAP and the field the current website relies on.

And yes, using #852 to track

Reply:
To clarify, the API converts `eventEndingTime` from CAP into just `end`. We don't know why they rename it, but we've confirmed that's what happens.

Reply:
Yes we chatted to them about it. It was originally built to support the OG weather.gov legacy site. It doesn't serve a purpose anymore.

We've been told to rely on `eventEndingTime`. If you all feel it's not a risk to rely on `end` knowing that, than proceed.

Reply:
We'll stick with `end` for now because it's a top-level property. We have to dig deeper into the data structure to get to `eventEndingTime` and the only difference between the two is that one of them is in UTC and the other includes a UTC offset, but they report the exact same time.